### PR TITLE
[Develop] Applying new Hpopt

### PR DESCRIPTION
**Pull request by eunwoosh**
### Summary

- This PR applies new Hpopt.
- SMBO isn't supported yet.
- This PR also includes unit test of `hpo.py`.

### Detail
I reimplemented hpopt to execute ASHA as explained in original paper and also refactor codes.
To apply this "new hpopt", OTX part also needs to be modified and this is for that.
I also added a feature that best hpo weight is now re-used when training a model after HPO.

### Purpose of code modification
**-> otx/algorithms/anomaly/adapters/anomalib/callbacks/progress.py**
Current HPO gets progress iteration information from score, but new HPO gets that from progress argument. So unnecessary code is removed.

**-> otx/algorithms/common/utils/callback.py**
For same reason as above, unnecessary code is removed.

**-> otx/cli/tools/train.py**
This is for code refactoring. Now HPO changes hyper parameters by optimized ones in `environment` variable.

**-> otx/cli/utils/tests.py**
Modify test code to align with new Hpopt.

**-> otx/cli/utils/hpo.py**
Code is modified to align with new hpopt. Because all code is changes, so recommend not to compare it with previous code when code review.

**-> otx/mpa/modules/hooks/adaptive_training_hooks.py**
A model is evaluated at least once before training is done.

**-> otx/mpa/modules/hooks/checkpoint_hook.py**
This file is in charge of saving a classification model weight. Current classification task saves a best model weight named "best.pth". But HPO needs both latest model weight and best model weight, so code is modified to save both.

**-> otx/mpa/modules/hooks/save_initial_weight_hook.py**
Code refactoring.

**-> requirements/base.txt**
Use "new_hpo" branch in hpopt repository. hpopt will be consolidated into OTX, so it's used for a while.

**-> tests/unit/cli/utils/test_hpo.py**
Unit test of "hpo.py"

### CODE REVIEW

File: tests/unit/cli/utils/test_hpo.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    check_hpopt_available
+)
+
+from tempfile import TemporaryDirectory
+
+CLASSIFCATION_TASK = {TaskType.CLASSIFICATION}
+DETECTION_TASK = {TaskType.DETECTION, TaskType.INSTANCE_SEGMENTATION, TaskType.ROTATED_DETECTION}
+SEGMENTATION_TASK = {TaskType.SEGMENTATION}
+ANOMALY_TASK = {TaskType.ANOMALY_CLASSIFICATION, TaskType.ANOMALY_DETECTION, TaskType.ANOMALY_SEGMENTATION}
+MPA_TASK = CLASSIFCATION_TASK | DETECTION_TASK | SEGMENTATION_TASK
```
Comment (harimkang) [inline tests/unit/cli/utils/test_hpo.py]: We understand that we have agreed to stop using the word `MPA` now :)

File: requirements/base.txt [hunk 2]
```diff
@@ -17,6 +17,6 @@ torch>=1.8.1, <=1.9.0
 torchvision>=0.9.1, <=0.10.1
 scikit-learn==0.24.*
 Shapely>=1.7.1,<=1.8.0
-hpopt==0.3.0
+hpopt @ git+https://github.com/openvinotoolkit/hyper_parameter_optimization@new_asha_with_refactored_hpopt
```
Comment (harimkang) [inline requirements/base.txt]: Is this a temporary measure for CI testing?

File: tests/unit/cli/utils/test_hpo.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+        assert not check_hpopt_available()
+
+def test_run_hpo():
+    pass
+
+def test_get_best_hpo_weight():
+    pass
+
+def test_run_trial():
+    pass
```
Comment (harimkang) [inline tests/unit/cli/utils/test_hpo.py]: Need to check pre-commit test

**Review by harimkang — COMMENTED** [review]
Added some comments!

File: requirements/base.txt (hunk 2 above)
Comment (eunwoosh) [reply requirements/base.txt]: I will also consolidate hpopt into OTX after merging this PR to reduce size of PR. So it's a just transition period.

File: otx/cli/utils/hpo.py [hunk 4]
```diff
@@ -70,766 +56,827 @@ def _check_hpo_enabled_task(task_type: TaskType):
     ]
 
 
-def _is_mpa_framework_task(task_type: TaskType):
-    """Check whether mpa framework task."""
-    return _is_cls_framework_task(task_type) or _is_det_framework_task(task_type) or _is_seg_framework_task(task_type)
+def check_hpopt_available() -> bool:
+    """Check whether hpopt is avaiable."""
```
Comment (JihwanEom) [inline otx/cli/utils/hpo.py:60]:
```suggestion
    """Check whether hpopt is available."""
```

File: otx/cli/utils/hpo.py [hunk 5]
```diff
@@ -70,766 +56,827 @@ def _check_hpo_enabled_task(task_type: TaskType):
...
+        """
+        return self._task_type in [
+            TaskType.ANOMALY_CLASSIFICATION,
+            TaskType.ANOMALY_DETECTION,
+            TaskType.ANOMALY_SEGMENTATION,
+        ]
 
-    task = task_class(task_environment=environment, output_path=args.work_dir)
+    def get_batch_size_name(self) -> str:
+        """Give an proper batch size name depending on frameowrk.
```
Comment (JihwanEom) [inline otx/cli/utils/hpo.py]:
```suggestion
        """Give a proper batch size name depending on framework.
```

File: otx/cli/utils/hpo.py [hunk 6]
```diff
@@ -70,766 +56,827 @@ def _check_hpo_enabled_task(task_type: TaskType):
...
 
-    if hpopt_cfg.get("resume", False):
-        task.set_resume_path_to_config(hpo_weight_path)  # prepare finetune stage to resume
+        return batch_size_name
 
-    if args.load_weights:
-        assert environment.model is not None
-        environment.model.configuration.configurable_parameters = hyper_parameters
+    def get_epoch_name(self) -> str:
+        """Give an proper epoch name depending on frameowrk.
```
Comment (JihwanEom) [inline otx/cli/utils/hpo.py]:
```suggestion
        """Give a proper epoch name depending on framework.
```

File: otx/cli/utils/hpo.py [hunk 7]
```diff
@@ -70,766 +56,827 @@ def _check_hpo_enabled_task(task_type: TaskType):
...
+        learning_parameters = self._convert_parameter_group_to_dict(learning_parameters)
+        hyper_parameter = {f"learning_parameters.{key}": val for key, val in learning_parameters.items()}
+        return hyper_parameter
 
-    return hpopt_cfg
+    def _convert_parameter_group_to_dict(self, parameter_group) -> Dict[str, Any]:
+        """Convert parameter group to dictionary.
 
+        Args:
+            parameter_group : parameter gruop
```
Comment (JihwanEom) [inline otx/cli/utils/hpo.py:270]:
```suggestion
            parameter_group : parameter group
```

File: otx/cli/tools/train.py [hunk 8]
```diff
@@ -218,12 +218,9 @@ def main():  # pylint: disable=too-many-branches
         )
 
     if args.enable_hpo:
-        task = run_hpo(args, environment, dataset, template.task_type)
-        if task is None:
-            print("cannot run HPO for this task. will train a model without HPO.")
-            task = task_class(task_environment=environment, output_path=args.work_dir)
-    else:
-        task = task_class(task_environment=environment, output_path=args.work_dir)
+        run_hpo(args, environment, dataset, data_roots)
```
Comment (sungmanc) [inline otx/cli/tools/train.py]: If this function changes the environment, need to add a description for notifying that this function will change the environment.

i.e. ```modified_environment =  run_hpo()```

**Comment by eunwoosh** [comment]
All pre-merge TCs are passed in local.
hpo_premerge

File: otx/cli/utils/hpo.py [hunk 9]
```diff
@@ -70,766 +56,827 @@ def _check_hpo_enabled_task(task_type: TaskType):
     ]
 
 
-def _is_mpa_framework_task(task_type: TaskType):
-    """Check whether mpa framework task."""
-    return _is_cls_framework_task(task_type) or _is_det_framework_task(task_type) or _is_seg_framework_task(task_type)
+def check_hpopt_available() -> bool:
```
Comment (sungmanc) [inline otx/cli/utils/hpo.py]: How about using `is_hpopt_available()`

File: otx/cli/utils/hpo.py [hunk 10]
```diff
@@ -70,766 +56,827 @@ def _check_hpo_enabled_task(task_type: TaskType):
...
+        return self.is_cls_framework_task() or self.is_det_framework_task() or self.is_seg_framework_task()
 
-def _is_anomaly_framework_task(task_type: TaskType):
-    """Check whether anomaly framework task."""
-    return task_type in [
-        TaskType.ANOMALY_CLASSIFICATION,
-        TaskType.ANOMALY_DETECTION,
-        TaskType.ANOMALY_SEGMENTATION,
-    ]
+    def is_cls_framework_task(self) -> bool:
```
Comment (sungmanc) [inline otx/cli/utils/hpo.py:87]: This is the design choice, maybe we can gathering `is_cls_framework_task` `is_det_framework_task` `is_seg_framework_task`, ... to `get_task_type`?
``` python
get_task_type(task_type:TaskType):
    if task_type == TaskType.CLASSIFICATION:
         return 
```

File: tests/unit/cli/utils/test_hpo.py [hunk 11]
```diff
@@ -0,0 +1,684 @@
...
+DETECTION_TASK = {TaskType.DETECTION, TaskType.INSTANCE_SEGMENTATION, TaskType.ROTATED_DETECTION}
+SEGMENTATION_TASK = {TaskType.SEGMENTATION}
+ANOMALY_TASK = {TaskType.ANOMALY_CLASSIFICATION, TaskType.ANOMALY_DETECTION, TaskType.ANOMALY_SEGMENTATION}
+MMCV_TASK = CLASSIFCATION_TASK | DETECTION_TASK | SEGMENTATION_TASK
+ALL_TASK = MMCV_TASK | ANOMALY_TASK
+OTX_ROOT_PATH = Path(otx.__file__).parent
+
+
+class TestTaskManager:
+    @pytest.mark.parametrize("task", MMCV_TASK)
```
Comment (harimkang) [inline tests/unit/cli/utils/test_hpo.py:43]: If it's a unit test, could you add a unit test decorator `@e2e_pytest_unit`? 

from tests.test_suite.e2e_test_system import e2e_pytest_unit

Comment (eunwoosh) [reply tests/unit/cli/utils/test_hpo.py:43]: Ah, ok. I'll update it.

File: otx/cli/tools/train.py (hunk 8 above)
Comment (eunwoosh) [reply otx/cli/tools/train.py]: I agree. I'll update it.

File: otx/cli/utils/hpo.py (hunk 9 above)
Comment (eunwoosh) [reply otx/cli/utils/hpo.py]: yours is better, I think. I'll update it.

File: otx/cli/utils/hpo.py (hunk 10 above)
Comment (eunwoosh) [reply otx/cli/utils/hpo.py:87]: Actually, I think current implementation is easier to read. How about let it be?

Comment (sungmanc) [reply otx/cli/utils/hpo.py:87]: No problems :)

File: requirements/detection.txt [hunk 12]
```diff
@@ -2,4 +2,4 @@ mmcv-full==1.6.1
 mmdet==2.25.1
 pytorchcv
 mmcls==0.23.2
-timm==0.6.12
```
Comment (JihwanEom) [inline requirements/detection.txt]: Do we need rebase?

Comment (eunwoosh) [reply requirements/detection.txt]: Thanks for check! I'll remove it.

**Review by JihwanEom — DISMISSED** [review]

**Review by harimkang — DISMISSED** [review]
If there is no problem in the CI test, it seems to be okay to merge.

**Review by JihwanEom — APPROVED** [review]

**Review by supersoob — APPROVED** [review]

**Pull request merged**